src/openassetio-python/cmodule/src/errors/exceptionsBinding.cpp

```diff
@@ -11,87 +14,376 @@
-      PyErr_SetObject(pyClass.ptr(), pyInstance.ptr());
-    };
+/**
+ * Container for a list of types and their corresponding string
+ * identifiers.
+ *
+ * @tparam Type List of types.
+ */
+template <class... Type>
+struct TypesAndIds {
+  static constexpr std::size_t kSize = sizeof...(Type);
+  using Types = std::tuple<Type...>;
+  const std::array<std::string_view, kSize> ids;
+};
 
-    // Handle the different possible C++ exceptions, creating the
-    // corresponding Python exception and setting it as the active
-    // exception in this thread.
-    try {
-      std::rethrow_exception(std::move(pexc));
-    } catch (const ConfigurationException &exc) {
-      setPyException("ConfigurationException", exc);
-    } catch (const InputValidationException &exc) {
-      setPyException("InputValidationException", exc);
-    } catch (const NotImplementedException &exc) {
-      setPyException("NotImplementedException", exc);
-    } catch (const UnhandledException &exc) {
-      setPyException("UnhandledException", exc);
-    } catch (const OpenAssetIOException &exc) {
-      setPyException("OpenAssetIOException", exc);
-    }
-  });
+/**
+ * Exhaustive list of all OpenAssetIO-specific C++ exception types and
+ * their corresponding Python class names.
+ *
+ * Note base classes must come _after_ subclasses (asserted at
+ * compile-time, below). This is so that try-catch blocks can be ordered
+ * such that more-derived exceptions come before less-derived. See
+ * `tryCatch`.
+ */
+constexpr auto kCppExceptionsAndPyClassNames =
+    TypesAndIds<BatchElementException, NotImplementedException, UnhandledException,
+                ConfigurationException, InputValidationException, OpenAssetIOException>{
+        "BatchElementException",  "NotImplementedException",  "UnhandledException",
+        "ConfigurationException", "InputValidationException", "OpenAssetIOException"};
+
+/**
+ * Convenience struct deriving compile-time properties from the above
+ * list of OpenAssetIO-specific exception class types and names.
+ *
+ * In particular there are two lists, such that given a single index,
+ * the C++ exception class or the Python class name can be queried at
+ * compile time.
+ */
+struct CppExceptionsAndPyClassNames {
+  using Type = decltype(kCppExceptionsAndPyClassNames);
+  /// "Array" of C++ exception classes.
+  template <std::size_t I>
+  using Exceptions = std::tuple_element_t<I, typename Type::Types>;
+  /// Array of Python exception class names.
+  static constexpr std::array kClassNames = kCppExceptionsAndPyClassNames.ids;
+  /// Total number of exceptions in the list(s).
+  static constexpr std::size_t kSize = Type::kSize;
+  /**
+   * Convenience for a sequence of indices, used for compile-time
+   * iteration over C++ exception classes.
+   */
+  static constexpr auto kIndices = std::make_index_sequence<kSize>{};
+};
+
+/**
+ * @}
+ */
+
+/**
+ * @section asserts Compile-time assertions.
```

Comment:
I'm not keen on that, since the `exceptionsValidation.hpp` header on its own would have lots of red squiggles in my editor (since it depends on stuff defined "above").

I am keen on finding a way to hide some of this assertion logic though, since I totally agree it causes issues with readability.

Reply:
Yeah that's a fair point.